### PR TITLE
Testing multiple versions of Rails

### LINKED ISSUE #273: Rewrite tests to work on Rails 4.0, 4.1, 4.2, and 5.0
Right now the tests for bootstrap_form are hardcoded to work with Rails 4.0.0. As mentioned in #229, the tests break if one upgrades the Gemfile.lock to Rails 4.1 or newer. I think improving our testing is extremely important and should be our top priority before pursuing other features (e.g. Bootstrap 4 support).

Ideally our tests should pass on all supported versions of Rails, meaning 4.0, 4.1, 4.2, and 5.0.

Does anyone have suggestions for how we can accomplish this?

I will gladly accept a PR if you want to take a shot. Please make a proposal in this thread first before submitting a PR.

### PR DESCRIPTION
**Pull request by koenpunt**
This fixes #273

### CODE REVIEW

**Comment by koenpunt** [comment]
And I'm not sure why the Travis build isn't being triggered for this PR..?

**Comment by mattbrictson** [comment]
Thanks for taking a first shot at this. I'll try to review later today.

I would like to sort out the Travis issue, but unfortunately I'm a not a project owner, just a contributor. That means I cannot see or edit the Travis config.

@potenza @carloslopes Can you help us turn on Travis for PRs?

**Comment by mattbrictson** [comment]
This is an excellent start! I like the use of the appraisal gem, and it is hooked up nicely with Travis.

Clearly we have some work to do in order to get tests passing on 4.1+.

My concerns so far are:

1. How do we do this in a piecemeal fashion so that we can merge into master sooner rather than later, but at the same time keep Travis green? Maybe work on getting 4.0 and 4.1 passing first, merge, then fix 4.2 and 5.0 in a separate PR?
2. The output from rails-dom-testing when an assertion fails is very hard to interpret (it seems to just call `#inspect` on the DOM nodes). Can we do better in terms of the output for failing DOM assertions?

```
 25) Failure:
BootstrapFormTest#test_the_field_is_wrapped_with_div.field_with_errors_when_form_for_is_used [/home/travis/build/bootstrap-ruby/rails-bootstrap-forms/test/bootstrap_form_test.rb:175]:
--- expected
+++ actual
@@ -1 +1 @@
-#<HTML::Node:0xXXXXXX @parent=nil, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Node:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[], @line=1, @position=137, @name="input", @attributes={"name"=>"utf8", "type"=>"hidden", "value"=>"&#x2713;"}, @closing=:self>], @line=1, @position=90, @name="div", @attributes={"style"=>"margin:0;padding:0;display:inline"}, @closing=nil>, #<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Text:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[], @line=1, @position=315, @content="Email">], @line=1, @position=260, @name="label", @attributes={"class"=>"control-label required", "for"=>"user_email"}, @closing=nil>], @line=1, @position=229, @name="div", @attributes={"class"=>"field_with_errors"}, @closing=nil>, #<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[], @line=1, @position=365, @name="input", @attributes={"class"=>"form-control", "id"=>"user_email", "name"=>"user[email]", "type"=>"text"}, @closing=:self>], @line=1, @position=334, @name="div", @attributes={"class"=>"field_with_errors"}, @closing=nil>, #<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Text:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[], @line=1, @position=473, @content="can&#39;t be blank, is too short (minimum is 5 characters)">], @line=1, @position=448, @name="span", @attributes={"class"=>"help-block"}, @closing=nil>], @line=1, @position=195, @name="div", @attributes={"class"=>"form-group has-error"}, @closing=nil>], @line=1, @position=0, @name="form", @attributes={"accept-charset"=>"UTF-8", "action"=>"/users", "class"=>"new_user", "id"=>"new_user", "method"=>"post"}, @closing=nil>], @line=0, @position=0>
+#<HTML::Node:0xXXXXXX @parent=nil, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Node:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[], @line=1, @position=116, @name="input", @attributes={"name"=>"utf8", "type"=>"hidden", "value"=>"&#x2713;"}, @closing=:self>], @line=1, @position=90, @name="div", @attributes={"style"=>"display:none"}, @closing=nil>, #<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Text:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[], @line=1, @position=294, @content="Your e-mail address">], @line=1, @position=239, @name="label", @attributes={"class"=>"control-label required", "for"=>"user_email"}, @closing=nil>], @line=1, @position=208, @name="div", @attributes={"class"=>"field_with_errors"}, @closing=nil>, #<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[], @line=1, @position=358, @name="input", @attributes={"class"=>"form-control", "id"=>"user_email", "name"=>"user[email]", "type"=>"text"}, @closing=:self>], @line=1, @position=327, @name="div", @attributes={"class"=>"field_with_errors"}, @closing=nil>, #<HTML::Tag:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[#<HTML::Text:0xXXXXXX @parent=#<HTML::Tag:0xXXXXXX ...>, @children=[], @line=1, @position=466, @content="can&#39;t be blank, is too short (minimum is 5 characters)">], @line=1, @position=441, @name="span", @attributes={"class"=>"help-block"}, @closing=nil>], @line=1, @position=174, @name="div", @attributes={"class"=>"form-group has-error"}, @closing=nil>], @line=1, @position=0, @name="form", @attributes={"accept-charset"=>"UTF-8", "action"=>"/users", "class"=>"new_user", "id"=>"new_user", "method"=>"post"}, @closing=nil>], @line=0, @position=0>
```

**Comment by koenpunt** [comment]
We might be better of with a custom parser, which only checks a limited set of attributes. Because I already see issues with certain versions of Rails outputting different attributes and additional elements than others

**Comment by mattbrictson** [comment]
> We might be better of with a custom parser

Rather than risk reinventing the wheel, maybe we can continue using rails-dom-parser, but instead of `assert_dom_equal`, use `assert_select` and test for certain CSS expressions?

BTW, thanks for taking the lead on this. If the scope of the work gets overwhelming and you want to split it up somehow, let me know and I'll do my best to help out.

**Comment by mattbrictson** [comment]
Also: I think Travis integration is working now. Next time you push commits to this PR the Travis build status should get hooked up.

**Comment by koenpunt** [comment]
Cool, will see if I can squeeze in a update somewhere this week

**Comment by koenpunt** [comment]
@mattbrictson I finally got around fixing the tests (with the `equivalent-xml` gem proposed by @lcreid).

I can imagine that certain logic that is now part of `assert_equivalent_xml` should be added to specific tests instead, but please have a look and let me know what you think :)

File: test/bootstrap_radio_button_test.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 require 'test_helper'
+require 'nokogiri'
+require 'equivalent-xml'
```
Comment (mattbrictson) [inline test/bootstrap_radio_button_test.rb]: I think these `require` statements should move to `test_helper.rb`?

**Comment by mattbrictson** [comment]
@lcreid I tried to invite you as a reviewer to this PR, but apparently GH only allows maintainers (?) to be reviewers. In any case, I would appreciate your 2¢ before merging this if you have time. Thanks!

File: test/bootstrap_radio_button_test.rb (hunk 1 above)
Comment (lcreid) [inline test/bootstrap_radio_button_test.rb]: The only reason those `require` statements are there is that's the file I arbitrarily chose to initially do the work. What I checked in yesterday I hadn't even reviewed myself. I didn't realize how quickly you were going to take the commit. :-) I agree the `require`s should go in `test_helper.rb`.

**Comment by lcreid** [comment]
@mattbrictson it looks good to me. @koenpunt obviously knows this issue inside out, to be able to add the special cases so quickly. Let me know if I can help out in any other way.

**Comment by koenpunt** [comment]
I've updated per the requested changes. We might consider enabling bundler cache to speed up the build on travis?

**Comment by mattbrictson** [comment]
@koenpunt Looks great. I don't think we need the bundler cache as part of this PR; we can always add it separately.

Did you want a chance to squash some of the commits? I am fine either way. I'll merge as soon as I get a thumbs-up from you.

**Comment by koenpunt** [comment]
> Did you want a chance to squash some of the commits?

No I'm good :shipit:

**Comment by mattbrictson** [comment]
Merged! Thanks for the awesome PR ✨

**Pull request merged**